File: linetrace/linetrace/TurnCalc.java
package linetrace;

public class TurnCalc {

	private float KP = 0.0F;
	private float KD = 0.0F;

	private float BrightTarget = 0.0F;

	private BrightSensor bright;

	private float prediff = 0.0F;

	public TurnCalc(BrightSensor bright){
		this.bright = bright;
	}

	public float getTurn() {

		float curdiff = BrightTarget - bright.getBright();

		float proportion = KP * curdiff;

		float differential = KD * (curdiff - prediff);

		prediff = curdiff;

		return (int)(proportion+differential);
	}

	public void setTarget(float white, float black) {
		this.BrightTarget = (white+black)/2.0F;
	}

}

File: linetrace/EV3RT/linetrace/TurnCalc.cpp
#include "ETBrightSensor.h"
#include "TurnCalc.h"

namespace linetrace
{

TurnCalc::TurnCalc(ETBrightSensor* bright, float KP, float KD)
:BrightTarget(0.0F), prediff(0.0F)
{
	this->bright = bright;
	this->KP = KP;
	this->KD = KD;
}

float TurnCalc::getTurn()
{
	float curdiff = BrightTarget - this->bright->getBright();

	float turn = (KP * curdiff) + (KD * (curdiff - prediff));

	prediff = curdiff;

	return turn;
}

void TurnCalc::setTarget(float white, float black)
{
	BrightTarget = (white + black) / 2.0;
}
}  // namespace linetrace
